extractor.go

```diff
@@ -93,52 +159,66 @@ func (e *Extractor) Extract(root, rel string) error {
 		// TODO(lwsanty): infinite loop detection
 		return e.Extract(root, dst)
 	})
+	if e.num == 1 {
+		return err
+	}
+	close(jobs)
+	if err != nil {
+		return err
+	}
+	wg.Wait()
+	select {
+	case err = <-errc:
+		return err
+	default:
+	}
+	return nil
 }
 
 // ExtractFrom extracts imports from a given file content, assuming it had a given path.
 //
 // The path is used for language detection only. It won't access any files locally and won't fetch dependency manifests
 // as Extract may do.
 func (e *Extractor) ExtractFrom(path string, content []byte) (*File, error) {
-	lang := enry.GetLanguage(path, content)
-	f := &File{Path: path, Lang: lang}
-	if lang == enry.OtherLanguage {
-		// unknown language - skip
-		return f, nil
-	}
-	l := LanguageByName(lang)
-	if l == nil {
-		// emit the file-language mapping; no imports
-		return f, nil
-	}
-	// import extraction
-	list, err := l.Imports(content)
-	if err != nil {
-		return f, err
+	return Extract(path, content)
+}
+
+func (e *Extractor) worker(jobs <-chan *extractJob, errc chan<- error) {
+	buf := make([]byte, e.maxSize)
+	for args := range jobs {
+		args.buf = buf
+		if err := e.processFile(*args); err != nil {
+			errc <- err
+			return
+		}
 	}
-	sort.Strings(list)
-	f.Imports = list
-	return f, nil
 }
 
-func (e *Extractor) processFile(fname, path string, sample bool) error {
-	f, err := os.Open(path)
+func (e *Extractor) processFile(args extractJob) error {
+	if len(args.buf) == 0 {
+		panic("buffer must be set")
+	}
+	f, err := os.Open(args.path)
 	if err != nil {
 		return err
 	}
 	defer f.Close()
-	var data []byte
-	if sample {
-		data = make([]byte, e.maxSize)
-		_, err = io.ReadFull(f, data)
-	} else {
-		data, err = ioutil.ReadAll(f)
-	}
-	if err != nil {
-		return err
+
+	data := args.buf
+	total := 0
+	for left := data; len(left) > 0; {
+		n, err := f.Read(left)
+		total += n
+		left = left[n:]
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
 	}
 	_ = f.Close()
-	return e.processAndEmit(fname, path, data)
+	data = data[:total]
```

Comment:
What does it do - sampling?
Maybe it's better here to use `https://golang.org/pkg/io/#LimitReader`

Reply:
Yes, it enforces sampling.

Good idea! I changed it to `LimitReader` as well as `bytes.Buffer` to simplify the code a bit.